procrastinate/worker.py

```diff
@@ -15,70 +15,109 @@ def __init__(
+                message = "Stop requested, no job to finish"
+
+            self.logger.info(message, extra=extra)
+
+        queues = self.queues
         if queues:
-            queue_names = f"queues {', '.join(queues)}"
+            queues_display = f"queues {', '.join(queues)}"
         else:
-            queue_names = "all queues"
+            queues_display = "all queues"
 
         self.logger.info(
-            f"Starting worker on queues {queue_names}",
+            f"Starting worker on queues {queues_display}",
             extra={"action": "start_worker", "queues": queues},
         )
-        with signals.on_stop(self.stop):
-            while not self.stop_requested:
-
-                try:
-                    await self.process_jobs_once()
-                except exceptions.StopRequested:
-                    break
-                except exceptions.NoMoreJobs:
-                    self.logger.debug(
-                        f"Waiting for new jobs on queues {queue_names}",
-                        extra={"action": "waiting_for_jobs", "queues": queues},
-                    )
 
-                await self.job_store.wait_for_jobs()
+        with signals.on_stop(stop):
+            await self.single_worker(notify_event=notify_event)
 
         self.logger.info(
-            f"Stopped worker on {queue_names}",
+            f"Stopped worker on {queues_display}",
             extra={"action": "stop_worker", "queues": queues},
         )
 
-    async def process_jobs_once(self) -> NoReturn:
-        while True:
-            # We only leave this loop with an exception:
-            # either NoMoreJobs or StopRequested
-            await self.process_next_job()
-
-    async def process_next_job(self) -> None:
-        log_context = self.log_context
-        job = await self.job_store.fetch_job(self.queues)
-        if job is None:
-            raise exceptions.NoMoreJobs
+    async def single_worker(self, notify_event: asyncio.Event):
+        queues = self.queues
+        if queues:
+            queues_display = f"queues {', '.join(queues)}"
+        else:
+            queues_display = "all queues"
+
+        while not self.stop_requested:
+            job = await self.job_store.fetch_job(self.queues)
+            if not job:
+                self.logger.debug(
+                    f"Waiting for new jobs on queues {queues_display}",
+                    extra={"action": "waiting_for_jobs", "queues": queues},
+                )
+                notify_event.clear()
+                await asyncio.wait([notify_event.wait()], timeout=self.timeout)
+                notify_event.clear()
+                continue
+            await self.process_job(job=job)
+
+    async def process_job(self, job: jobs.Job) -> None:
+        job_context = job.get_context()
+        self.current_job_context = job_context
+        log_context: types.JSONDict = {
+            "worker_name": self.worker_name,
+            "job": job_context,
+        }
```

Comment:
Previously the `log_context` had no `worker_name` key when `name` wasn't set or was `None` in the worker constructor. Now there always is a `worker_name` key, whose value might be `None`.